### ISSUE TITLE
[no-unnecessary-condition] Many false positives with `undefined`

### ISSUE THREAD

**Issue by Timmmm** [typescript-eslint/typescript-eslint#2511]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
const a = [1, 2, 3];
if (a[4] !== undefined) {
  console.log(a[4]);
}
```

**Expected Result**

No errors.

**Actual Result**

It says `a[4] !== undefined` is an unnecessary condition.

**Additional Info**

This is because [Typescript's type signatures for array access is wrong](https://github.com/Microsoft/TypeScript/issues/13778). It should include `undefined` as a possible return value but it does not.

Until they fix that, can I suggest that an option is added to this rule to ignore unnecessary conditions involving `undefined`?

**Versions**

All latest.

**Comment by bradzacher** [comment]
the rule currently only supports ignoring array checks using `if (arr[i])`.
Happy to accept a PR to extend support to also include `if (arr[i] !== undefined)`

**Comment by danielnixon** [comment]
Totally unsolicited, but @Timmmm you might like https://github.com/danielnixon/total-functions#get-type-safe-array-index-operator

IMO it's better to use something like that than to carve out exceptions in `no-unnecessary-condition`.

**Comment by ark120202** [comment]
Note that there is an open PR to TypeScript implementing an option adding `undefined` to index signatures: https://github.com/microsoft/TypeScript/pull/39560

**Comment by Timmmm** [comment]
Thanks for the hint @danielnixon. I did consider something like that but I think I would struggle to convince my coworkers (they don't care about type correctness as much as me). Also it is kind of harder to read. :-/

**Comment by danielnixon** [comment]
> I would struggle to convince my coworkers

We'll never solve that sort of social problem with technology, but if you can manage to persuade your team to turn on an eslint rule like https://github.com/danielnixon/eslint-plugin-total-functions#total-functionsno-unsafe-subscript then at least they'll have someone else (the linter) reminding them who isn't you...

> Also it is kind of harder to read. :-/

That's fair. Let's hope https://github.com/microsoft/TypeScript/pull/39560 lands soon.

**Comment by Timmmm** [comment]
Ok how about something like [this](https://github.com/Timmmm/typescript-eslint/commit/aa92737c50fd2f2323230dc1a3c088b2e3902e83)? I have only tested it by hacking the Javascript in my `node_modules`, but it seems to work.

If you think it is a reasonable solution I will update it with documentation & tests (assuming there are tests).

**Comment by bradzacher** [comment]
I don't think that there should be a new option. Sorry I tagged it wrong.

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L266

We already have some checks in the rule around ignoring array things. It would be worth investigating and figuring out how those work and if we can make them apply to the case of `!== undefined`

**Comment by Timmmm** [comment]
I think that only works if the expression in the condition is an array index expression, and if you check where that is called, it only applies to the `??` operator. In other words it knows not to complain about this:

```
const a = [1, 2, 3];
const _b = a[4] ?? 0;
```

But it complains about this:

```
const c = a[5];
const _d = c ?? 0;
```

![image](https://user-images.githubusercontent.com/376842/92709179-9b7ed280-f34e-11ea-93c0-914977653b60.png)

And it complains about this:

```
if (arr[10] === undefined) {
  ...
}
```

Arguably that should be fixed by relaxing the `??` condition, but then it would still complain about this:

```
const x = arr[10];
if (x === undefined) {
  ...
}
```

I presume the kind of analysis you'd need to do to figure out "did this expression originate from an array access?" is too complex for ESLint.

> I don't think that there should be a new option. 

So do you think I should just remove the `undefined` checks entirely?

**Comment by bradzacher** [comment]
The two are separate issues here:

The first issue is that it's not detecting this case:

```ts
const a = [1, 2, 3];
if (a[4] !== undefined) {
}
```

It should. The rule should very easily be able to detect this case. There's probably a reason that it wasn't added originally, but it should really ignore this.



The second issue is that the rule doesn't currently follow variable references.
Which means that it cannot understand this case:

```ts
const a = [1,2,3];
const b = a[4];
if (b) {}
```

It really should. It just hasn't yet because doing so adds complexity to the rule.

----

> I presume the kind of analysis you'd need to do to figure out "did this expression originate from an array access?" is too complex for ESLint.

For some cases, yes. For most cases - no.

ESLint supports scope analysis. It's very easy to follow a variable reference within a scope.
Certain cases are hard to deal with, but anything involving `const` is dead easy.

```ts
const a = [1,2,3];

const b = a[1];
if (b) {} // easy to follow

const c = b;
if (c) {} // harder to follow, but still possible

let d = a[1];
if (d) {} // still easy to follow as the variable has 1 assignment

let e = 0;
e = a[2];
if (e) {} // nigh impossible to follow, as you don't know exactly what the value of e is at this moment
// it would need some control-flow analysis to determine.

let f = 0;
function foo() { f = a[2] }
if (f) {} // just illustrating why the above case quickly devolves to be impossible

// the same applies for `var` instead of `let`

test(b);
function test(arg: number) {
  if (arg) {} // impossible to follow to the actual arg type.
}


/// a.ts
export const g = a[1];
/// b.ts
import { g } from './a'
if (g) {} // impossible to follow to the actual type
```

**Comment by Timmmm** [comment]
Ah ok, if it is possible to use scope analysis to do this properly then that would be good. Way beyond my ability to hack on the code though, so would you accept my solution as a temporary measure until someone more experienced with the ESLint code can add full scope analysis?

**Comment by bradzacher** [comment]
Happy to accept a PR which just updates the current logic to also apply our existing array check to `a[0] !== undefined`.

**Comment by Timmmm** [comment]
That would definitely be useful, although based on my brief survey of our codebase the vast majority of array access cases do not look like that - usually they are:

```
let foo = a[0];
if (foo === undefined) {
```

Looking through our codebase, the check did find several misuses of `=== undefined` so it would be really nice to avoid just ignoring those checks entirely. Maybe I'll have a go at this scope analysis thing. What's the most similar rule that does scope analysis?

**Comment by bradzacher** [comment]
`consistent-type-imports` is the only rule we have in our package which leverage scope analysis.

You can also review the eslint core codebase, just search the codebase for `context.getScope`

**Comment by Timmmm** [comment]
Actually I just saw [this comment](https://github.com/microsoft/TypeScript/issues/13778#issuecomment-691331991) that we'll be able to fix the array access type so it returns `| undefined` in Typescript 4.1. Kind of looked like it was going to take a lot longer than that to land! I think fixing the underlying issue in Typescript is a better solution than working around the current broken behaviour in ESLint so I will just wait for that and close this.

**Comment by boris-petrov** [comment]
What about the same issue with `undefined` not with arrays but with objects?

```typescript
const x: { [key: string]: number } = { a: 1 };
if (x.b != null) {
}
```

This triggers a warning that the null-check is not needed but it is. The merged PR in TypeScript works only for arrays as far as I understand. What are we to do with such cases then? Explicitly allowing `undefined` like so - `{ [key: string]: number | undefined }` is cumbersome and also annoying in other places where I have to put `!` where I know I have a value.
